File: mini-era/cv/svm/src/matlab/getAlphaFromTrainSet.m
function [a,b,C,d,dim,e,eps,a_result,b_result,X,tolerance,Y,ret]=getAlphaFromTrainSet(N, d16trn_1, d16trn_2, Iterations)


tolerance=0.001;
C=0.05;
d = -1;
dim=256;
eps=0.001;
a_result=zeros(Iterations,N);
b_result=zeros(Iterations,1);
ret = 0;
Y = 0;

X=usps_read_partial(d16trn_1, d16trn_2,1,1,N/Iterations, Iterations);

for iter=1:Iterations
    Y=usps_read_partial(d16trn_1,d16trn_2,iter,0,N/Iterations, Iterations);
    if iter==1
    fWriteMatrix(Y, './');
    end

    a=zeros(N,1);
    b=0;
    e=zeros(N,1);
    ExamineAll=1;  
    cnt=0;
    NumChanged=0;
    while (NumChanged>0 || ExamineAll == 1)
        cnt=cnt+1;
        NumChanged=0;
        if ExamineAll==1
            for i=1:N
                [a,b,e,ret] = examineExample(i, a, b, C, e, X, Y, tolerance, N, eps, dim);
                NumChanged=NumChanged+ret;
            end
        else
            for i=1:N
                if a(i,1) > 0 && a(i,1) < C
                    [a,b,e,ret] = examineExample(i, a, b, C, e, X, Y, tolerance, N, eps, dim);
                    NumChanged=NumChanged+ret;
                end
            end
        end
        if ExamineAll==1
            ExamineAll=0;
        elseif NumChanged==0 
            ExamineAll=1;
        end
    end
    
    for r = 1:N
        a_result(iter,r) = a(r,1);
    end
    
%    a_result=transpose(a); %KVS: Problem using transpose function here. So wrote the code above
    b_result(iter,1)=b;
end

File: mini-era/cv/svm/src/c/getAlphaFromTrainSet.c
/********************************
Author: Sravanthi Kota Venkata
********************************/

#include "svm.h"

alphaRet* getAlphaFromTrainSet(int N, F2D* trn1, F2D* trn2, int iterations)
{
    float tolerance, C, eps, *b;
    F2D *a_result, *b_result;
    int NumChanged, r, ExamineAll, cnt, d, dim, ret, iter, i;
    F2D *X, *Y;
    F2D *a, *e;

    b = malloc(sizeof(float));
    alphaRet* alpha;
    alpha = (alphaRet*)malloc(sizeof(alphaRet));
    tolerance = 0.001;
    C = 0.05;
    d = -1;
    dim = 256;
    eps = 0.001;
    a_result = fSetArray(iterations, N, 0);
    b_result = fSetArray(iterations, 1, 0);
    ret = 0;
    
    X = usps_read_partial( trn1, trn2, 0, 1, (N/iterations), iterations);
    
    for(iter=0; iter<iterations; iter++)
    {
        Y = usps_read_partial( trn1, trn2, iter, 0, N/iterations, iterations);
        
        a = fSetArray(N, 1, 0);
        arrayref(b,0) = 0;                  /** check if ptr **/
        e = fSetArray(N, 1, 0);
        ExamineAll = 1;
        cnt = 0;
        NumChanged = 0;

        while(NumChanged>0 || ExamineAll == 1)
        {
            cnt = cnt + 1;
            NumChanged = 0;
            if(ExamineAll == 1)
            {
                for(i=0; i<N; i++)
                {
                    ret = examineExample(i, a, b, C, e, X, Y, tolerance, N, eps, dim);
                    NumChanged = NumChanged + ret;
                }
            }
            else
            {
                for(i=0; i<N; i++)
                {
                    if( asubsref(a,i) > 0 && asubsref(a,i) <C )
                    {
                        ret = examineExample(i, a, b, C, e, X, Y, tolerance, N, eps, dim);
                        NumChanged = NumChanged + ret;
                    }
                }
            }
            if(ExamineAll == 1)
                ExamineAll = 0;
            else if(NumChanged == 0)
                ExamineAll = 1;
        }

        for(r=0; r<N; r++)
            subsref(a_result,iter,r) = asubsref(a,r);   /** a_result has size iteration,N .. Check **/
        asubsref(b_result,iter) = arrayref(b,0);

        fFreeHandle(Y);
        fFreeHandle(e);
        fFreeHandle(a);
    }
  
    alpha->C = C;
    alpha->d = d;
    alpha->dim = dim;
    alpha->eps = eps;
    alpha->a_result = a_result;
    alpha->b_result = b_result;
    alpha->a = a;
    alpha->b = arrayref(b,0);
    alpha->X = X;
    alpha->tolerance = tolerance;
    alpha->ret;
    
    free(b);
   
    return alpha; 

}




